frontend/ui/menusorter.lua

```diff
@@ -0,0 +1,193 @@
+                -- regular, just insert a menu action
+                else
+                    if order_number_id == "----------------------------" then
+                        -- it's a separator
+                        tmp_menu_table[order_number] = self.separator
+                    elseif item_table[order_number_id] ~= nil then
+                        item_table[order_number_id].id = order_number_id
+                        tmp_menu_table[order_number] = item_table[order_number_id]
+                        -- remove reference from item_table so it won't show up as orphaned
+                        item_table[order_number_id] = nil
+                    end
+                end
+
+            end
+            -- compress menus
+            -- if menu_items were missing we might have a table with gaps
+            -- but ipairs doesn't like that and quits when it hits nil
+            local k, v
+            local previous_key
+            local new_index = 1
+            k, v = next(tmp_menu_table, nil)
+            while k do
+                if type(k) == "number" then
+                    if v.id == "----------------------------" then
+                        new_index = new_index - 1
+                        self.menu_table[order_id][new_index].separator = true
+                    else
+                        -- fix the index
+                        self.menu_table[order_id][new_index] = tmp_menu_table[k]
+                    end
+
+                    previous_key = k
+                    new_index = new_index + 1
+                end
+                k, v = next(tmp_menu_table, k)
+            end
+        else
+            DEBUG("menu id not found:", order_id)
+        end
+    end
+
+    -- now do the submenus
+    DEBUG("SUBMENUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUS")
+    DEBUG("self.sub_menus", sub_menus)
+    for i,sub_menu in ipairs(sub_menus) do
+        local sub_menu_position = self:findById(self.menu_table["KOMenu:menu_buttons"], sub_menu) or nil
+        if sub_menu_position and sub_menu_position.id then
+            sub_menu_position.sub_item_table = self.menu_table[sub_menu]
+            -- remove reference from top level output
+            self.menu_table[sub_menu] = nil
+            -- remove reference from input so it won't show up as orphaned
+            item_table[sub_menu] = nil
+        end
+    end
+    -- @TODO avoid this extra mini-loop
+    -- cleanup, top-level items shouldn't have sub_item_table
+    for i,top_menu in ipairs(self.menu_table["KOMenu:menu_buttons"]) do
+        self.menu_table["KOMenu:menu_buttons"][i] = self.menu_table["KOMenu:menu_buttons"][i].sub_item_table
+    end
+    
+    -- remove top level reference before orphan handling
+    item_table["KOMenu:menu_buttons"] = nil
+        --attach orphans based on menu_hint
+    DEBUG("ORPHAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAANSS", util.tableSize(item_table))
```

Comment:
This should use the new logger functionality. And of course I'll tone down the debug statements. :-P